src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1140,6 +1230,123 @@ private void IssueBuildSubmissionToScheduler(BuildSubmission submission, bool al
+                            .ToList();
+                        foreach (var node in unblockedNodes)
+                        {
+                            var targetList = targetLists[node];
+                            if (targetList.Count == 0)
+                            {
+                                // An empty target list here means "no targets" instead of "default targets", so don't even build it.
+                                finishedNodes.Add(node);
+                                blockedNodes.Remove(node);
+
+                                waitHandle.Set();
+
+                                continue;
+                            }
+
+                            var request = new BuildRequestData(node.ProjectInstance, targetList.ToArray());
+
+                            // TODO Tack onto the existing submission instead of pending a whole new submission for every node
+                            // Among other things, this makes BuildParameters.DetailedSummary produce a summary for each node, which is not desirable.
+                            var innerBuildSubmission = PendBuildRequest(request);
+                            buildingNodes.Add(innerBuildSubmission, node);
+                            blockedNodes.Remove(node);
+                            innerBuildSubmission.ExecuteAsync(finishedBuildSubmission =>
+                            {
+                                ProjectGraphNode finishedNode;
+                                lock (graphBuildStateLock)
+                                {
+                                    finishedNode = buildingNodes[finishedBuildSubmission];
+
+                                    finishedNodes.Add(finishedNode);
+                                    buildingNodes.Remove(finishedBuildSubmission);
+                                }
+
+                                if (finishedNode == entryNode)
+                                {
+                                    // The overall submission is complete, so report it as complete
+                                    submission.CompleteLogging(true);
+                                    ReportResultsToSubmission(new BuildResult(submission.BuildRequest, finishedBuildSubmission.BuildResult, null));
+                                }
+
+                                waitHandle.Set();
+                            }, null);
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                if (ExceptionHandling.IsCriticalException(ex))
+                {
+                    throw;
+                }
+
+                // ProjectGraph throws an aggregate exception with InvalidProjectFileException inside when evaluation fails
+                if (ex is AggregateException aggregateException && aggregateException.InnerExceptions.All(innerException => innerException is InvalidProjectFileException))
```

Comment:
It'll be hard to test this e2e until #3873 is merged.

Reply:
Yeah, let's block this on that. Looks like it should go in soon. @yash256 fyi.